File: assignments/week-5/0287-find-the-duplicate-number/solution.java
class Solution {
    public int findDuplicate(int[] nums) {
        // Binary Search Solution: On Answer
        // Logic: Counting the numbers <= mid if count
        // is greater than the mid then our ans is mid &
        // go to left else go to right;
        // the reason why our answer is mid cause you tell me
        // how many elements are gonna be <= 3 or any number n?
        // the count should be not more than n right? for 3 there
        // could be just 1, 2, and 3 at max but we found the count
        // more than 3 elements then don't you think we have a repeated
        // number as one of them: could be 1, 2, and 3;
        // so we leverage this idea and try binary Search on answer;

        int n = nums.length;

        int left = 1;
        int right = n;

        while(left <= right){
            int mid = left + (right - left) / 2;
            int elsCount = 0;

            for(int num : nums){
                if(num <= mid){
                    elsCount += 1;
                }
            }

            if(elsCount <= mid){
                left = mid + 1;
            }
            else{
                right = mid - 1;
            }
        }

        return left;
    }
}

File: assignments/week-5/0287-find-the-duplicate-number/solution.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        '''
        # Counting Sort Solution
        n = len(nums)
        idx_i = 0

        while idx_i < n:
            curr_num = nums[idx_i]

            if (curr_num - 1) != idx_i:
                if curr_num == nums[curr_num - 1]:
                    return curr_num
                else:
                    nums[idx_i], nums[curr_num - 1] = nums[curr_num - 1], nums[idx_i]
            else:
                idx_i += 1
        
        '''

        '''
        # Bit Manipulation

        duplicate_num = 0
        n = len(nums)

        for val in range(1, n):
            duplicate_num = duplicate_num ^ val
        
        for num in nums:
            duplicate_num = duplicate_num ^ num

        return duplicate_num
        # 3, 3, 3, 3, 3 => tricky case; 
        '''

        # slow = 0
        # fast = 0
        # slow += 1
        # fast += 2

        # 1, 3, 4, 2, 2
        # slow = 1, 3, 2
        # fast = 1, 2, 2

        n = len(nums)
        slow = nums[0]
        fast = nums[0]

        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]

            if slow == fast:
                break

        slow = nums[0]
        # slow2 = 1, 3, 2
        # slow = 2, 4, 2

        while slow != fast:
            slow = nums[slow]
            fast = nums[fast]

        return slow